################################################################################
SHORT DESCRIPTION: 
################################################################################
Re-create the driver stack as the path may have changed on ring resume.

################################################################################
LONG DESCRIPTION: 
################################################################################
The original issue for this is unknown but the patch seems reasonable given
what it says it is fixing.

Keeping this patch and possibly upstreaming it if the original problem
can be reproduced.

Not sure what the mercurial change set references...

HG changeset patch
Parent 98865fdac09e36bc2c6af3c6d516d3d3fa144307

################################################################################
CHANGELOG 
################################################################################

################################################################################
REMOVAL 
################################################################################

################################################################################
UPSTREAM PLAN
################################################################################

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
Index: xen-4.10.0/tools/blktap2/drivers/tapdisk-vbd.c
===================================================================
--- xen-4.10.0.orig/tools/blktap2/drivers/tapdisk-vbd.c
+++ xen-4.10.0/tools/blktap2/drivers/tapdisk-vbd.c
@@ -441,6 +441,20 @@ fail:
 	return err;
 }
 
+static void
+free_driver_stack(td_vbd_t *vbd)
+{
+	td_vbd_driver_info_t *driver;
+
+	while(!list_empty(&vbd->driver_stack)) {
+		driver = list_entry(vbd->driver_stack.next,
+		    td_vbd_driver_info_t, next);
+		list_del(&driver->next);
+		free(driver->params);
+		free(driver);
+	}
+}
+
 /* this populates a vbd type based on path */
 int
 tapdisk_vbd_parse_stack(td_vbd_t *vbd, const char *path)
@@ -1648,6 +1662,14 @@ tapdisk_vbd_resume_ring(td_vbd_t *vbd)
 		goto out;
 	}
 
+	/* re-create the driver stack as the path may have changed */
+	free_driver_stack(vbd);
+	err = tapdisk_vbd_parse_stack(vbd, message);
+	if (err) {
+		err = -EINVAL;
+		goto out;
+	}
+
 	tapdisk_vbd_start_queue(vbd);
 
 	for (i = 0; i < TD_VBD_EIO_RETRIES; i++) {
